sdk/core/azure-core/azure/core/tracing/_inference_api_instrumentor_impl.py

```diff
@@ -0,0 +1,379 @@
+
+_inference_traces_enabled: bool = False
+_trace_inference_content: bool = False
+INFERENCE_GEN_AI_SYSTEM_NAME = "azure.ai.inference"
+
+class TraceType(str, Enum):
+    """An enumeration class to represent different types of traces."""
+
+    INFERENCE = "Inference"
+
+
+def _set_attributes(span: AbstractSpan, *attrs: tuple[str, Any]) -> None:
+    for attr in attrs:
+        key, value = attr
+        if value is not None:
+            span.add_attribute(key, value)
+
+
+def _add_request_chat_message_event(span: AbstractSpan, **kwargs: Any) -> None:
+    for message in kwargs.get("messages", []):
+        try:
+            message = message.as_dict()
+        except AttributeError:
+            pass
+
+        if message.get("role"):
+            name = f"gen_ai.{message.get('role')}.message"
+            span.span_instance.add_event(
+                name=name,
+                attributes={
+                    "get_ai.system": INFERENCE_GEN_AI_SYSTEM_NAME,
+                    "gen_ai.event.content": json.dumps(message)
+                }
+            )
+
+
+def _add_request_chat_attributes(span: AbstractSpan, **kwargs: Any) -> None:
+    _set_attributes(
+        span,
+        ("gen_ai.system", INFERENCE_GEN_AI_SYSTEM_NAME),
+        ("gen_ai.request.model", kwargs.get("model")),
+        ("gen_ai.request.max_tokens", kwargs.get("max_tokens")),
+        ("gen_ai.request.temperature", kwargs.get("temperature")),
+        ("gen_ai.request.top_p", kwargs.get("top_p")),
+    )
+
+
+def _add_response_chat_message_event(span: AbstractSpan, result: _models.ChatCompletions) -> None:
+    global _trace_inference_content
+    for choice in result.choices:
+        if _trace_inference_content:
+            response: dict[str, Any] = {
+                "message": {"content": choice.message.content},
+                "finish_reason": str(choice.finish_reason),
+                "index": choice.index,
+            }
+            attributes={
+                "get_ai.system": INFERENCE_GEN_AI_SYSTEM_NAME,
+                "gen_ai.event.content": json.dumps(response)
+            }
+        else:
+            response: dict[str, Any] = {
+                "finish_reason": str(choice.finish_reason),
+                "index": choice.index,
+            }
+            attributes={
+                "get_ai.system": INFERENCE_GEN_AI_SYSTEM_NAME,
+            }
+        if choice.message.tool_calls:
+            response["message"]["tool_calls"] = [tool.as_dict() for tool in choice.message.tool_calls]
+        span.span_instance.add_event(name="gen_ai.choice", attributes=attributes)
+
+
+def _add_response_chat_attributes(span: AbstractSpan, result: _models.ChatCompletions | _models.StreamingChatCompletionsUpdate) -> None:
+    _set_attributes(
+        span,
+        ("gen_ai.response.id", result.id),
+        ("gen_ai.response.model", result.model),
+        ("gen_ai.response.finish_reason", str(result.choices[-1].finish_reason)),
```

Comment:
There will be exception in streaming scenario, where result.choices can be None

Reply:
This seems to work
("gen_ai.response.finish_reason", getattr(result.choices[-1], "finish_reason", None) if hasattr(result, "choices") and len(result.choices) > 0 else None),

Reply:
I suggest to fallback to `error` if we didn't get a finish reason. 

I.e. if there was one choice and it didn't complete, the `gen_ai.response.finish_reasons` should be `["error"]`
If there were two and the first was received fully and another was not, it would be `gen_ai.response.finish_reason = ["stop", "error"]`

WDYT?